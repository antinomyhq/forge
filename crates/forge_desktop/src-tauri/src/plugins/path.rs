use std::path::PathBuf;

use tauri::plugin::{Builder, TauriPlugin};
use tauri::Runtime;

#[tauri::command]
async fn get_home_dir() -> Result<Option<String>, String> {
    Ok(dirs::home_dir().map(|p| p.to_string_lossy().to_string()))
}

#[tauri::command]
async fn get_config_dir() -> Result<Option<String>, String> {
    Ok(dirs::config_dir().map(|p| p.to_string_lossy().to_string()))
}

#[tauri::command]
async fn join_paths(base: &str, components: Vec<&str>) -> Result<String, String> {
    let mut path = PathBuf::from(base);
    for component in components {
        path = path.join(component);
    }
    Ok(path.to_string_lossy().to_string())
}

pub fn init<R: Runtime>() -> TauriPlugin<R> {
    Builder::new("path")
        .invoke_handler(tauri::generate_handler![
            get_home_dir,
            get_config_dir,
            join_paths
        ])
        .build()
}
