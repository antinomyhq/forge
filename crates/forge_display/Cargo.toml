[package]
name = "forge_display"
version = "0.1.0"
edition = "2021"

[dependencies]
derive_setters.workspace = true
colored.workspace = true
chrono.workspace = true
similar.workspace = true
console.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
pretty_assertions.workspace = true
strip-ansi-escapes.workspace = true