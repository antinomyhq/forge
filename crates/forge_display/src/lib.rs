pub mod diff;
pub mod grep;
pub mod title;

pub use diff::DiffFormat;
pub use grep::GrepFormat;
pub use title::TitleFormat;
