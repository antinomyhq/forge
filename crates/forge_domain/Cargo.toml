[package]
name = "forge_domain"
version = "0.1.0"
edition = "2021"

[dependencies]
handlebars = "6.0.0"
Inflector = "0.11.4"
anyhow = "1.0"
async-trait = "0.1.84"
chrono = { version = "0.4.39", features = ["serde"] }
derive_more = { version = "1.0.0", features = ["from", "display"] }
derive_setters = "0.1.6"
futures = "0.3.31"
html-escape = "0.2.13"
nom = "8.0.0"
quick-xml = { version = "0.37.2", features = ["serialize"] }
schemars = "0.8.21"
serde = "1.0.217"
serde_json = "1.0.134"
strum = "0.26.3"
strum_macros = "0.26.4"
thiserror = "2.0.11"
tokio = { version = "1.42.0", features = ["full", "test-util"] }
tokio-stream = "0.1.17"
uuid = { version = "1.11.0", features = [
    "v4",
    "fast-rng",
    "macro-diagnostics",
    "serde",
] }

[dev-dependencies]
insta = { version = "1.34.0", features = ["json"] }
pretty_assertions = "1.4.1"
