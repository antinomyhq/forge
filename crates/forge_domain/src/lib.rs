mod agent;
mod arena;
mod chat_request;
mod chat_response;
mod chat_stream_ext;
mod command;
mod config;
mod context;
mod context_ext;
mod conversation;
mod env;
mod errata;
mod error;
mod file;
mod learning;
mod message;
mod model;
mod provider;
mod stream_ext;
mod tool;
mod tool_call;
mod tool_call_parser;
mod tool_choice;
mod tool_definition;
mod tool_name;
mod tool_result;
mod tool_service;
mod tool_usage;
mod user_interaction;
mod workflow;

pub use agent::*;
pub use chat_request::*;
pub use chat_response::*;
pub use chat_stream_ext::*;
pub use command::*;
pub use config::*;
pub use context::*;
pub use context_ext::*;
pub use conversation::*;
pub use env::*;
pub use errata::*;
pub use error::*;
pub use file::*;
pub use learning::*;
pub use message::*;
pub use model::*;
pub use provider::*;
pub use stream_ext::*;
pub use tool::*;
pub use tool_call::*;
pub use tool_call_parser::*;
pub use tool_choice::*;
pub use tool_definition::*;
pub use tool_name::*;
pub use tool_result::*;
pub use tool_service::*;
pub use tool_usage::*;
pub use user_interaction::*;
pub use workflow::*;

/// Core domain trait providing access to services and repositories.
/// This trait follows clean architecture principles for dependency management
/// and service/repository composition.
pub trait ForgeDomain {
    /// The concrete type implementing file read service capabilities
    type FileReadService: file::FileReadService;
    /// The concrete type implementing tool service capabilities
    type ToolService: ToolService;
    /// The concrete type implementing provider service capabilities
    type ProviderService: ProviderService;
    /// The concrete type implementing conversation repository
    type ConversationRepo: ConversationRepository;
    /// The concrete type implementing configuration repository
    type ConfigRepo: ConfigRepository;
    /// The concrete type implementing environment repository
    type EnvironmentRepo: EnvironmentRepository;

    /// Get a reference to the tool service instance
    fn tool_service(&self) -> &Self::ToolService;
    /// Get a reference to the provider service instance
    fn provider_service(&self) -> &Self::ProviderService;
    /// Get a reference to the conversation repository instance
    fn conversation_repository(&self) -> &Self::ConversationRepo;
    /// Get a reference to the configuration repository instance
    fn config_repository(&self) -> &Self::ConfigRepo;
    /// Get a reference to the environment repository instance
    fn environment_repository(&self) -> &Self::EnvironmentRepo;

    /// Get a reference to the file read service instance
    fn file_read_service(&self) -> &Self::FileReadService;
}
