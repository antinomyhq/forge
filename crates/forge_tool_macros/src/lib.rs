use proc_macro::TokenStream;
use proc_macro2::TokenTree;
use quote::{quote, ToTokens};
use syn::{parse_macro_input, DeriveInput};

#[proc_macro_derive(ToolDescription)]
pub fn derive_description(input: TokenStream) -> TokenStream {
    // Parse the input struct or enum
    let input = parse_macro_input!(input as DeriveInput);
    let name = &input.ident;
    let generics = &input.generics;

    // Collect doc lines from all `#[doc = "..."]` attributes
    let mut doc_lines = Vec::new();
    for attr in &input.attrs {
        // Check if the attribute is `#[doc(...)]`
        if attr.path().is_ident("doc") {
            // `parse_nested_meta` calls the provided closure for each nested token
            // of the attribute (e.g., = "some doc string").

            for t in attr
                .to_token_stream()
                .into_iter()
                .filter_map(|t| match t {
                    TokenTree::Group(lit) => Some(lit.stream()),
                    _ => None,
                })
                .flatten()
            {
                if let TokenTree::Literal(lit) = t {
                    let str = lit.to_string();
                    // Remove surrounding quotes from the doc string
                    let clean_str = str.trim_matches('"').to_string();
                    if !clean_str.is_empty() {
                        doc_lines.push(clean_str);
                    }
                }
            }
        }
    }

    // Join all lines with a space (or newline, if you prefer)
    if doc_lines.is_empty() {
        panic!("No doc comment found for {}", name);
    }
    let doc_string = doc_lines.join("\n").trim().to_string();

    // Generate an implementation of `ToolDescription` that returns the doc string
    let expanded = if generics.params.is_empty() {
        quote! {
            impl ToolDescription for #name {
                fn description(&self) -> String {
                    #doc_string.into()
                }
            }
        }
    } else {
        quote! {
            impl #generics ToolDescription for #name #generics {
                fn description(&self) -> String {
                    #doc_string.into()
                }
            }
        }
    };

    expanded.into()
}
