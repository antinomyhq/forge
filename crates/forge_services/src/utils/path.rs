use std::path::{Path, PathBuf};

use anyhow::bail;

pub async fn get_git_root(from_dir: &Path) -> Option<PathBuf> {
    let output = tokio::process::Command::new("git")
        .args(["rev-parse", "--show-toplevel"])
        .current_dir(from_dir)
        .output()
        .await
        .ok()?;

    if output.status.success() {
        String::from_utf8(output.stdout)
            .ok()
            .map(|root| PathBuf::from(root.trim()))
    } else {
        None
    }
}

/// Ensures that the given path is absolute
///
/// # Arguments
/// * `path` - The path to validate
///
/// # Returns
/// * `Ok(())` if the path is absolute
/// * `Err(String)` with an error message if the path is relative
pub fn assert_absolute_path(path: &Path) -> anyhow::Result<()> {
    if !path.is_absolute() {
        bail!("Path must be absolute. Please provide an absolute path starting with '/' (Unix) or 'C:\\' (Windows)".to_string())
    } else {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_unix_absolute_path() {
        let path = Path::new("/absolute/path");
        assert!(assert_absolute_path(path).is_ok());
    }

    #[test]
    #[cfg(target_os = "windows")]
    fn test_windows_absolute_path() {
        let path = Path::new("C:\\Windows\\Path");
        assert!(assert_absolute_path(path).is_ok());
    }

    #[test]
    fn test_basic_relative_path() {
        let path = Path::new("relative/path");
        assert!(assert_absolute_path(path).is_err());
    }

    #[test]
    fn test_current_dir_relative_path() {
        let path = Path::new("./current/path");
        assert!(assert_absolute_path(path).is_err());
    }

    #[test]
    fn test_parent_dir_relative_path() {
        let path = Path::new("../parent/path");
        assert!(assert_absolute_path(path).is_err());
    }
}
