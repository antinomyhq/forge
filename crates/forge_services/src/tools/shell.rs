use std::path::PathBuf;
use std::sync::Arc;

use anyhow::bail;
use forge_display::TitleFormat;
use forge_domain::{
    CommandOutput, Environment, EnvironmentService, ExecutableTool, NamedTool, ToolCallContext,
    ToolDescription, ToolName,
};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use strip_ansi_escapes::strip;

use crate::{CommandExecutorService, Infrastructure};

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
pub struct ShellInput {
    /// The shell command to execute.
    pub command: String,
    /// The working directory where the command should be executed.
    pub cwd: PathBuf,
    /// Whether to preserve ANSI escape codes in the output.
    /// If true, ANSI escape codes will be preserved in the output.
    /// If false (default), ANSI escape codes will be stripped from the output.
    #[serde(default)]
    pub keep_ansi: bool,
}

// Strips out the ansi codes from content.
fn strip_ansi(content: String) -> String {
    String::from_utf8_lossy(&strip(content.as_bytes())).into_owned()
}

/// Formats command output by wrapping non-empty stdout/stderr in XML tags.
/// stderr is commonly used for warnings and progress info, so success is
/// determined by exit status, not stderr presence. Returns Ok(output) on
/// success or Err(output) on failure, with a status message if both streams are
/// empty.
fn format_output(mut output: CommandOutput, keep_ansi: bool) -> anyhow::Result<String> {
    let mut formatted_output = String::new();

    if !keep_ansi {
        output.stderr = strip_ansi(output.stderr);
        output.stdout = strip_ansi(output.stdout);
    }

    if !output.stdout.trim().is_empty() {
        formatted_output.push_str(&format!("<stdout>{}</stdout>", output.stdout));
    }

    if !output.stderr.trim().is_empty() {
        if !formatted_output.is_empty() {
            formatted_output.push('\n');
        }
        formatted_output.push_str(&format!("<stderr>{}</stderr>", output.stderr));
    }

    let result = if formatted_output.is_empty() {
        if output.success {
            "Command executed successfully with no output.".to_string()
        } else {
            "Command failed with no output.".to_string()
        }
    } else {
        formatted_output
    };

    if output.success {
        Ok(result)
    } else {
        Err(anyhow::anyhow!(result))
    }
}

/// Executes shell commands with safety measures using restricted bash (rbash).
/// Prevents potentially harmful operations like absolute path execution and
/// directory changes. Use for file system interaction, running utilities,
/// installing packages, or executing build commands. For operations requiring
/// unrestricted access, advise users to run forge CLI with '-u' flag. Returns
/// complete output including stdout, stderr, and exit code for diagnostic
/// purposes.
#[derive(ToolDescription)]
pub struct Shell<I> {
    env: Environment,
    infra: Arc<I>,
}

impl<I: Infrastructure> Shell<I> {
    /// Create a new Shell with environment configuration
    pub fn new(infra: Arc<I>) -> Self {
        let env = infra.environment_service().get_environment();
        Self { env, infra }
    }
}

impl<I> NamedTool for Shell<I> {
    fn tool_name() -> ToolName {
        ToolName::new("forge_tool_process_shell")
    }
}

#[async_trait::async_trait]
impl<I: Infrastructure> ExecutableTool for Shell<I> {
    type Input = ShellInput;

    async fn call(&self, context: ToolCallContext, input: Self::Input) -> anyhow::Result<String> {
        // Validate empty command
        if input.command.trim().is_empty() {
            bail!("Command string is empty or contains only whitespace".to_string());
        }
        let title_format = TitleFormat::debug(format!("Execute [{}]", self.env.shell.as_str()))
            .sub_title(&input.command);

        context.send_text(title_format).await?;

        let output = self
            .infra
            .command_executor_service()
            .execute_command(input.command, input.cwd)
            .await?;

        format_output(output, input.keep_ansi)
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;
    use std::{env, fs};

    use pretty_assertions::assert_eq;

    use super::*;
    use crate::attachment::tests::MockInfrastructure;

    /// Platform-specific error message patterns for command not found errors
    #[cfg(target_os = "windows")]
    const COMMAND_NOT_FOUND_PATTERNS: [&str; 2] = [
        "is not recognized",             // cmd.exe
        "'non_existent_command' is not", // PowerShell
    ];

    #[cfg(target_family = "unix")]
    const COMMAND_NOT_FOUND_PATTERNS: [&str; 3] = [
        "command not found",               // bash/sh
        "non_existent_command: not found", // bash/sh (Alternative Unix error)
        "No such file or directory",       // Alternative Unix error
    ];

    #[tokio::test]
    async fn test_shell_echo() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo 'Hello, World!'".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        assert!(result.contains("Mock command executed successfully"));
    }

    #[tokio::test]
    async fn test_shell_stderr_with_success() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        // Use a command that writes to both stdout and stderr
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: if cfg!(target_os = "windows") {
                        "echo 'to stderr' 1>&2 && echo 'to stdout'".to_string()
                    } else {
                        "echo 'to stderr' >&2; echo 'to stdout'".to_string()
                    },
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert_eq!(
            result,
            "<stdout>to stdout\n</stdout>\n<stderr>to stderr\n</stderr>"
        );
    }

    #[tokio::test]
    async fn test_shell_both_streams() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo 'to stdout' && echo 'to stderr' >&2".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert_eq!(
            result,
            "<stdout>to stdout\n</stdout>\n<stderr>to stderr\n</stderr>"
        );
    }

    #[tokio::test]
    async fn test_shell_with_working_directory() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let temp_dir = fs::canonicalize(env::temp_dir()).unwrap();

        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: if cfg!(target_os = "windows") {
                        "cd".to_string()
                    } else {
                        "pwd".to_string()
                    },
                    cwd: temp_dir.clone(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        assert_eq!(result, format!("<stdout>{}\n</stdout>", temp_dir.display()));
    }

    #[tokio::test]
    async fn test_shell_invalid_command() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "non_existent_command".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await;

        assert!(result.is_err());
        let err = result.unwrap_err();

        // Check if any of the platform-specific patterns match
        let matches_pattern = COMMAND_NOT_FOUND_PATTERNS
            .iter()
            .any(|&pattern| err.to_string().contains(pattern));

        assert!(
            matches_pattern,
            "Error message '{err}' did not match any expected patterns for this platform: {COMMAND_NOT_FOUND_PATTERNS:?}"
        );
    }

    #[tokio::test]
    async fn test_shell_empty_command() {
        let infra = Arc::new(MockInfrastructure::new());
        let shell = Shell::new(infra);
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await;
        assert!(result.is_err());
        assert_eq!(
            result.unwrap_err().to_string(),
            "Command string is empty or contains only whitespace"
        );
    }

    #[tokio::test]
    async fn test_description() {
        assert!(
            Shell::new(Arc::new(MockInfrastructure::new()))
                .description()
                .len()
                > 100
        )
    }

    #[tokio::test]
    async fn test_shell_pwd() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let current_dir = env::current_dir().unwrap();
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: if cfg!(target_os = "windows") {
                        "cd".to_string()
                    } else {
                        "pwd".to_string()
                    },
                    cwd: current_dir.clone(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert_eq!(
            result,
            format!("<stdout>{}\n</stdout>", current_dir.display())
        );
    }

    #[tokio::test]
    async fn test_shell_multiple_commands() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo 'first' && echo 'second'".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();
        assert_eq!(result, format!("<stdout>first\nsecond\n</stdout>"));
    }

    #[tokio::test]
    async fn test_shell_empty_output() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "true".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert!(result.contains("executed successfully"));
        assert!(!result.contains("failed"));
    }

    #[tokio::test]
    async fn test_shell_whitespace_only_output() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo ''".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert!(result.contains("executed successfully"));
        assert!(!result.contains("failed"));
    }

    #[tokio::test]
    async fn test_shell_with_environment_variables() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: "echo $PATH".to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await
            .unwrap();

        assert!(!result.is_empty());
        assert!(!result.contains("Error:"));
    }

    #[tokio::test]
    async fn test_shell_full_path_command() {
        let shell = Shell::new(Arc::new(MockInfrastructure::new()));
        // Using a full path command which would be restricted in rbash
        let cmd = if cfg!(target_os = "windows") {
            r"C:\Windows\System32\whoami.exe"
        } else {
            "/bin/ls"
        };

        let result = shell
            .call(
                ToolCallContext::default(),
                ShellInput {
                    command: cmd.to_string(),
                    cwd: env::current_dir().unwrap(),
                    keep_ansi: true,
                },
            )
            .await;

        // In rbash, this would fail with a permission error
        // For our normal shell test, it should succeed
        assert!(
            result.is_ok(),
            "Full path commands should work in normal shell"
        );
    }

    #[test]
    fn test_format_output_ansi_handling() {
        // Test with keep_ansi = true (should preserve ANSI codes)
        let ansi_output = CommandOutput {
            stdout: "\x1b[32mSuccess\x1b[0m".to_string(),
            stderr: "\x1b[31mWarning\x1b[0m".to_string(),
            success: true,
        };
        let preserved = format_output(ansi_output, true).unwrap();
        assert_eq!(
            preserved,
            "<stdout>\x1b[32mSuccess\x1b[0m</stdout>\n<stderr>\x1b[31mWarning\x1b[0m</stderr>"
        );

        // Test with keep_ansi = false (should strip ANSI codes)
        let ansi_output = CommandOutput {
            stdout: "\x1b[32mSuccess\x1b[0m".to_string(),
            stderr: "\x1b[31mWarning\x1b[0m".to_string(),
            success: true,
        };
        let stripped = format_output(ansi_output, false).unwrap();
        assert_eq!(
            stripped,
            "<stdout>Success</stdout>\n<stderr>Warning</stderr>"
        );
    }
}
