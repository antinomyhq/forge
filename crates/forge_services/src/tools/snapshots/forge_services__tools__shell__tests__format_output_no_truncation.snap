---
source: crates/forge_services/src/tools/shell.rs
expression: large_result
snapshot_kind: text
---
---
command: echo
exit_code: 0
---
<stdout>
ABCDEFGHIJKLMNOPQRSTUVWXYZ
</stdout>
